### ISSUE TITLE
Shadows should be applied to the stroke of a shape that is filled and stroked 

### ISSUE THREAD

**Issue by lehni** [paperjs/paper.js#892]
Check [this sketch](http://sketch.paperjs.org/#S/VY89C8IwEIb/ypGlCqXUwSXiYv+A4OBgHdLkJKExkUvaDqX/3doPpLe9z/Nyx/XMiTcyzm41RqlZyqRXv9wKAmlIWoQzOOzgKqLOions+tLBOBJdROLQGuyyOaSzIaFMEzgc8nwhL2Nt4a0f6wmhShYcIvkaV1HZBrfmblTUHI7rmqCF8t2/L2SdbNTFNjTdLd2wP43/VISi/njjYmD88Ry+): The fat stroke makes the shadow disappear completely.

**Comment by lehni** [comment]
Hmm, but the problem is if we apply it to the stroke, the shadow will appear both sides of it... There is no easy solution to this, actually.

**Comment by sapics** [comment]
I think so too, there would be no easy solution.

One of the way would be controlling by user, shadowForFill or shadowForStroke.

Another way would be following.
1) only fill: draw fill with shadow
2) only stroke: draw stroke with shadow
3) fill + stroke: draw stroke with shadow and strokeColor='rgba(0,0,0,0)', draw fill without shadow, draw stroke without shadow

**Comment by lehni** [comment]
Good thinking! 1) and 2) is already how it is currently dealt with, but I didn't think of 3)... I shall give that a go!

**Comment by lehni** [comment]
Unfortunately 3) doesn't work. It looks like the alpha of the stroke color is applied to the shadow as well, so when it's 0, both disappear.

**Comment by sapics** [comment]
Oh, i see. I did not know it...
I would like to report other method, if i find :)

**Comment by sapics** [comment]
How about 3) ->
1. path.translate(new Point(-x, -y))
2. path.shadowOffset = path.shadowOffset.add(new Point(x, y))
3. draw stroke with shadow
3. path.translate(new Point(x, y))
3. path.shadowOffset = path.shadowOffset.add(new Point(-x, -y))
4. draw fill without shadow
5. draw stroke without shadow

x, y are enough large number to draw stroke outside canvas.
Of course, it is not necessary translate each path._segments, just ctx.traslate would be enough :)

**Comment by sapics** [comment]
Above looks bad English :)
I rewrite above by js-like style.

```javascript
ctx.save()
ctx.translate(-x, -y)
var shadowOffset = item.shadowOffset
item.shadowOffset = item.shadowOffset.add(new Point(x, y))
item.drawStroke(ctx)
item.shadowOffset = shadowOffset
ctx.restore()

item.drawFillWithoutShadow(ctx)

item.drawStrokeWithoutShadow(ctx)
```

**Comment by lehni** [comment]
Hmm I'm not sure I understand your idea.What are `x` and `y` exactly in this example?

**Comment by sapics** [comment]
Thank you for creating jsfiddle, sorry for not enough explanation.
I have forked your jsfiddle to https://jsfiddle.net/meqonrg7/

This method draw "stroke shadow" without "stroke".
To do it, I draw "stroke" at "outside of canvas".

The x and y is only for drawing "stroke shadow" without "stroke".
I guess that x and y would depend on following.
1. <canvas> size
2. totalMatrix in draw
3. strokeBounds of item

P.S. Thank you for cleaning up all issues!

**Comment by lehni** [comment]
That's an interesting thought! But isn't this better then?

https://jsfiddle.net/lehni/hxwvpvu0/1/

And you're welcome! It was about time, so much easier to keep track of the pending things now.

**Comment by sapics** [comment]
It looks easier, but it draws stroke twice.
Thus, if the strokeStyle has alpha as "rgba(0,0,100,0.2)", then stroke is looks like "rgba(0,0,100,0.4)".

**Comment by lehni** [comment]
Oh yes, of course, you're right! We could half its alpha value if it is < 1, and then draw it twice with it... But I think it would end up mixing the background in differently that way, so not really a solution either.

What I liked better about this one is that it only creates the path information once. I think in terms of performance, your example probably ends up drawing the stroke twice as well, even though it is outside of the canvas the firs time, because it needs it to create the shadow.

This is to be verified, I am not certain about how this is handled internally, and could differ quite a bit across implementations.

**Comment by sapics** [comment]
I like better too, for drawing path at once.
My approach is not only bad for perfomance, but also makes worse and complicate paperjs source code.
This is not high priority problem, thus, it might be good for waiting a better idea. :)

**Comment by lehni** [comment]
I wish canvas had a `fillAndStroke()` command which would solve this issue here and also respect opacity properly (the stroke not showing through the fill). This would help hugely with performance...

**Comment by sapics** [comment]
Yeah, it looks good for many issues!

**Comment by lehni** [comment]
I have made the proposal now over at whatwg. Let's see what happens: https://github.com/whatwg/html/issues/518

**Comment by sapics** [comment]
:+1: It is so cool proposal!
It would be better to solve these problems without tricky or low performance methods.

**Comment by sapics** [comment]
I have an another idea about this problem.

When the item is (fill+stroke and shadow) or (Group and shadow).

1) Using un-direct draw like blendMode != 'normal' or opacity < 1.
2) Set item shadow to null, temporary (which revert at 4)
3) At `BlendMode.process`, set original shadow style to `dstContext` before `dstContext.drawImage` or `dstContext.putImageData`
4) Revert item shadow

This method has an advantage than before, it is also possible to do 'Group shadow'

**Comment by sapics** [comment]
It needs little fix like https://github.com/paperjs/paper.js/commit/b6a1a5bb0b90a1e769335d6b5f8be73765066128

**Comment by lehni** [comment]
Yes of course that works, and we already do this when there is opacity and the item has a fill and a stroke, but it just isn't fast : )

**Comment by lehni** [comment]
The setting of `style.getShadowColor().alpha ` to 0 seems hackish. Is this required so the shadow is not applied to the contents of the image? I think it would be better to introduce a new property on `param` for that. But I am happy to look into that part.

**Comment by sapics** [comment]
Yes, it would be slow.
To cache created imagedata, and use it if there are no difference, it would be faster.
But, it needs additional huge memories... so it would be bad idea :)

It makes sense, changing `style.getShadowColor().alpha` is bad manner :)
Introducing a new property to `param` is better.
Sorry, I wrote that commit just for checking how easy to fix.
And, I know little about CompoundPath, thus, my commit might be bad.

**Comment by sapics** [comment]
Is it better to make a PR after fixing `style.getShadowColor().alpha`?
If you feel bad because of performance or CompoundPath and so on, let's just trash it, I try to find the other method :)

**Comment by sapics** [comment]
I am sorry, I find some problems in my commit. Please let trash it.

**Comment by lehni** [comment]
Could you describe the problems? I didn't have time to look into this yet, but think it generally makes sense.

**Comment by sapics** [comment]
Sure!

First:
In the first place, I could not set blur styles to group. :)
We need to fix style functions first, for setting blur styles to group.

Second:
I forgot mainCtx.save() and mainCtx.restore().
It would fix easily.

---

By the way, I find another advantage of this method.

Currently, the non-direct drawed canvas (from CanvasProvider) is created with StrokeBounds size, thus, blur shadow is cutted as follow.
http://sketch.paperjs.org/#S/ZZFhS8MwEIb/yhGRtqyUzCG4igjzq4JMv237kKbpGprmSnd1yNh/9zJd0RnIJZf3XvJcchBetUbk4q0xpGuRCo1lyD9UD52iGh7Amz288jZbGk3Kb52JD+t+7YEHYfdsKsphNZUyBQ6b9KwVSITt0m7roM+CPvutV9a5J3TY5xBdVXcyGpVdrUrcL9zA0q28OD5bCqd08+05JvdhCTNwa4aWIakGr8miB10zt1k448sX7i9OxgZCj1lxFtgYTyY6gWu4gUeIVFlGwHd57Fvloj8m7JS29MkW/VMus/mcq6cjr6F32xocKL4AOL2UTE7sIfzjC4f8GUVvVNOh9bQT+Wpz/AI=

On the other hand, this method create blur shadow from non-direct drawed canvas.
Thus, blur shadow has always drawed correctly!

**Comment by lehni** [comment]
Oh I see. Yeah the problem with style being propagated up to all children is still there, as we need the style transition first. I agree.

**Comment by sapics** [comment]
Sorry, I have not explained about "GROUP SHADOW" which I want to do.
I am not good at English, thus, it would be better to explain by drawing.

![shadow](https://cloud.githubusercontent.com/assets/11169954/13352352/ff3a51b8-dccf-11e5-8842-21a93aa806e5.png)

**Comment by sapics** [comment]
Oops, previously, I said "blur shadow has always drawed correctly" with this method.
Sorry, i made a mistake, not always.
This method cannot draw blur shadow correctly at following case.

When the group drawed with non-direct, and the children has blur shadow, then children blur shadow would be cutted because group's StrokeBounds dose not include children's blur shadow.

Thus, to fix this more precisely, I guess that we need to use other method like `getShadowBounds` to get item's shadow bounds.
